src/mongocrypt-marking.c

```diff
@@ -150,40 +222,371 @@ _mongocrypt_marking_cleanup (_mongocrypt_marking_t *marking)
+      }                                                                        \
+                                                                               \
+      mc_##Name##DerivedFromDataToken_t *fromDataToken =                       \
+         mc_##Name##DerivedFromDataToken_new (crypto, token, value, status);   \
+      mc_##Name##Token_destroy (token);                                        \
+      if (!fromDataToken) {                                                    \
+         return false;                                                         \
+      }                                                                        \
+                                                                               \
+      mc_##Name##DerivedFromDataTokenAndCounter_t *fromTokenAndCounter =       \
+         mc_##Name##DerivedFromDataTokenAndCounter_new (                       \
+            crypto, fromDataToken, counter, status);                           \
+      mc_##Name##DerivedFromDataToken_destroy (fromDataToken);                 \
+      if (!fromTokenAndCounter) {                                              \
+         return false;                                                         \
+      }                                                                        \
+                                                                               \
+      _mongocrypt_buffer_copy_to (                                             \
+         mc_##Name##DerivedFromDataTokenAndCounter_get (fromTokenAndCounter),  \
+         out);                                                                 \
+      mc_##Name##DerivedFromDataTokenAndCounter_destroy (fromTokenAndCounter); \
+                                                                               \
+      return true;                                                             \
+   }
+
+DERIVE_TOKEN_IMPL (EDC)
+DERIVE_TOKEN_IMPL (ESC)
+DERIVE_TOKEN_IMPL (ECC)
+
+#undef DERIVE_TOKEN_IMPL
+
+
+static bool
+_fle2_placeholder_aes_ctr_encrypt (_mongocrypt_key_broker_t *kb,
+                                   const _mongocrypt_buffer_t *key,
+                                   const _mongocrypt_buffer_t *in,
+                                   _mongocrypt_buffer_t *out,
                                    mongocrypt_status_t *status)
+{
+   _mongocrypt_crypto_t *crypto = kb->crypt->crypto;
+   _mongocrypt_buffer_t iv, ciphertext;
+   const uint32_t cipherlen = MONGOCRYPT_IV_LEN + in->len;
+   uint32_t written = 0;
+
+   _mongocrypt_buffer_init_size (out, cipherlen);
+
+   BSON_ASSERT (
+      _mongocrypt_buffer_from_subrange (&iv, out, 0, MONGOCRYPT_IV_LEN));
+   if (!_mongocrypt_random (crypto, &iv, MONGOCRYPT_IV_LEN, status)) {
+      return false;
+   }
+
+   BSON_ASSERT (_mongocrypt_buffer_from_subrange (
+      &ciphertext, out, MONGOCRYPT_IV_LEN, out->len - MONGOCRYPT_IV_LEN));
+   if (!_native_crypto_aes_256_ctr_encrypt (
```

Comment:
I suggest using `_mongocrypt_fle2_do_encryption`. That will eventually use the crypto hooks.